pyvptree/include/VPLevelPartition.hpp

```diff
@@ -0,0 +1,167 @@
+
+    VPLevelPartition() {
+        // Constructs an empty level
+
+        _radius = 0;
+        _indexStart = -1;
+        _indexEnd = -1;
+    }
+
+    virtual ~VPLevelPartition() { clear(); }
+
+    SerializedState serialize() const {
+
+        SerializedState state;
+        std::vector<const VPLevelPartition *> flatten_tree_state;
+
+        flatten_tree(this, flatten_tree_state);
+        // we need to reverse since we will pop elements in reverse order when deserializing
+        std::reverse(flatten_tree_state.begin(), flatten_tree_state.end());
+
+        size_t total_size = flatten_tree_state.size() * (2 * sizeof(int64_t) + sizeof(float));
+        state.reserve(total_size);
+
+        // reverse the tree state since we will push it in a stack for serializing
+        for (const VPLevelPartition *elem : flatten_tree_state) {
+            if (elem == nullptr) {
+                state.push((float)(0));
+                state.push((int64_t)(-1));
+                state.push((int64_t)(-1));
+                continue;
+            }
+
+            state.push((float)(elem->_radius));
+            state.push((int64_t)(elem->_indexStart));
+            state.push((int64_t)(elem->_indexEnd));
+        }
+
+        if (state.size() != total_size) {
+            throw new std::out_of_range("invalid serialization state, offsets dont match!");
+        }
+
+        state.buildChecksum();
+
+        return state;
+    }
+
+    void deserialize(const SerializedState &state) {
+        clear();
+        SerializedState state_copy(state);
+
+        VPLevelPartition *recovered = rebuild_from_state(state_copy);
+        if (recovered == nullptr) {
+            return;
+        }
+
+        _left = recovered->_left;
+        _right = recovered->_right;
+        _radius = recovered->_radius;
+        _indexStart = recovered->_indexStart;
+        _indexEnd = recovered->_indexEnd;
+    }
+
+    bool isEmpty() { return _indexStart == -1; }
+    int64_t start() { return _indexStart; }
+    int64_t end() { return _indexEnd; }
+    int64_t size() { return _indexEnd - _indexStart + 1; }
+    void setRadius(distance_type radius) { _radius = radius; }
+    distance_type radius() { return _radius; }
+
+    void setChild(VPLevelPartition<distance_type> *left, VPLevelPartition<distance_type> *right) {
+        _left = left;
+        _right = right;
+    }
+
+    VPLevelPartition *left() const { return _left; }
+    VPLevelPartition *right() const { return _right; }
+
+    void print_state() { print_state(this, 0); }
+
+    private:
+    void clear() {
+        if (_left != nullptr)
+            delete _left;
+
+        if (_right != nullptr)
+            delete _right;
+
+        _left = nullptr;
+        _right = nullptr;
+    }
+
+    void print_state(const VPLevelPartition *root, int level = 0) {
+        if (root == nullptr) {
+            return;
+        }
+        print_state(root->left(), level + 1);
+        print_state(root->right(), level + 1);
+    }
+
```

Comment:
Unused

Reply:
Fixed on 81a53ff4fbc34b9916d4bb0df767ccbb01506712
Completedly changed string serialization and added to python bindings `to_string` method. Check the PR text to see the output. I think its nice I could see if the tree is balanced or not. It is indeed quite well balanced.